### PR TITLE
Support fractional seconds

### LINKED ISSUE #110: Support fractional seconds
Transaction timestamps have a 1s resolution. Add support for factional seconds.

See https://www.rfc-editor.org/rfc/rfc8949.html#name-epoch-based-date-time

### PR DESCRIPTION
**Pull request by Denommus**
## Description

Add a f64 variant to Timestamp so that it supports fractional seconds

## Related Issue

Fixes #110 

## Testing

Tested encoding and decoding in a unit test

## Checklist:

- [X] I have read and followed the CONTRIBUTING guidelines for this project.
- [X] I have added or updated tests and they pass.
- [X] I have added or updated documentation and it is accurate.
- [X] I have noted any breaking changes in this module or downstream modules.

### CODE REVIEW

File: src/many-types/src/timestamp.rs [hunk 1]
```diff
@@ -0,0 +1,132 @@
+use many_error::ManyError;
+use minicbor::{
+    data::{Tag, Type},
+    decode,
+    encode::{self, Write},
+    Decode, Decoder, Encode, Encoder,
+};
```
Comment (hansl) [inline src/many-types/src/timestamp.rs:7]:
```suggestion
use minicbor::{decode, encode, Decode, Decoder, Encode, Encoder};
use minicbor::data::{Tag, Type};
use minicbor::encode::Write;
```

**Review by hansl — APPROVED** [review]
LGTM with one note. Can you make sure many-framework still works? Removing `Eq` is a breaking change.

File: src/many-types/src/timestamp.rs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        Ok(Self::Int(secs))
+    }
+
+    pub const fn new_f64(secs: f64) -> Result<Self, ManyError> {
+        Ok(Self::Float(secs))
+    }
+
+    pub fn from_system_time(t: std::time::SystemTime) -> Result<Self, ManyError> {
+        let d = t.duration_since(std::time::UNIX_EPOCH).map_err(|_| {
+            ManyError::unknown("duration value can not represent system time".to_string())
```
Comment (fmorency) [inline src/many-types/src/timestamp.rs:39]: You can remove `.to_string()`

File: src/many-types/src/timestamp.rs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        })?;
+        Ok(Self::Float(d.as_secs_f64()))
+    }
+
+    pub fn as_system_time(&self) -> Result<std::time::SystemTime, ManyError> {
+        std::time::UNIX_EPOCH
+            .checked_add(std::time::Duration::new(self.secs(), self.nanos()))
+            .ok_or_else(|| {
+                ManyError::unknown("duration value can not represent system time".to_string())
```
Comment (fmorency) [inline src/many-types/src/timestamp.rs:48]: You can remove `.to_string()`

File: src/many-types/src/timestamp.rs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            })
+    }
+
+    pub fn secs(&self) -> u64 {
+        match self {
+            Timestamp::Int(a) => *a,
+            Timestamp::Float(a) => *a as u64,
```
Comment (fmorency) [inline src/many-types/src/timestamp.rs:55]: Add a comment as to why this is safe do to.

File: src/many-types/src/timestamp.rs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    }
+}
+
+impl std::ops::Add<f64> for Timestamp {
+    type Output = Timestamp;
+
+    fn add(self, rhs: f64) -> Self::Output {
+        Timestamp::new_f64(self.secs_f64().add(&rhs)).unwrap()
+    }
+}
```
Comment (fmorency) [inline src/many-types/src/timestamp.rs:91]: We should probably handle `checked_add()` as well

File: src/many-types/src/timestamp.rs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    let timestamp = Timestamp::new(10).unwrap();
+    let encoded = minicbor::to_vec(timestamp).unwrap();
+    let decoded: Timestamp = minicbor::decode(&encoded).unwrap();
+    assert_eq!(decoded, timestamp);
+
+    let timestamp = Timestamp::new_f64(1.1).unwrap();
+    let encoded = minicbor::to_vec(timestamp).unwrap();
+    let decoded: Timestamp = minicbor::decode(&encoded).unwrap();
+    assert_eq!(decoded, timestamp);
+}
```
Comment (fmorency) [inline src/many-types/src/timestamp.rs:132]: While you're modifying this, could you add more test cases?

I want to test edge/limit cases and the `add()` behavior.

**Review by fmorency — CHANGES_REQUESTED** [review]
Some minor stuff. Thanks!

**Comment by Denommus** [comment]
Closed per discussion with @hansl

**Pull request closed without merging**